outline/shadowsocks/connectivity.go

```diff
@@ -0,0 +1,91 @@
+package shadowsocks
+
+import (
+	"net"
+	"strconv"
+	"time"
+
+	oss "github.com/Jigsaw-Code/outline-go-tun2socks/shadowsocks"
+	"github.com/Jigsaw-Code/outline-ss-server/shadowsocks"
+)
+
+// Outline error codes. Must be kept in sync with definitions in outline-client/cordova-plugin-outline/outlinePlugin.js
+const (
+	noError                     = 0
+	unexpected                  = 1
+	noVPNPermissions            = 2
+	authenticationFailure       = 3
+	udpConnectivity             = 4
+	unreachable                 = 5
+	vpnStartFailure             = 6
+	ilegalConfiguration         = 7
+	shadowsocksStartFailure     = 8
+	configureSystemProxyFailure = 9
+	noAdminPermissions          = 10
+	unsupportedRoutingTable     = 11
+	systemMisconfigured         = 12
+)
+
+const reachabilityTimeout = 10 * time.Second
+
+// CheckConnectivity determines whether the Shadowsocks proxy can relay TCP and UDP traffic under
+// the current network. Parallelizes the execution of TCP and UDP checks, selects the appropriate
+// error code to return accounting for transient network failures.
+// Returns an error if an unexpected error ocurrs.
+func CheckConnectivity(host string, port int, password, cipher string) (int, error) {
+	client, err := shadowsocks.NewClient(host, port, password, cipher)
+	if err != nil {
+		// TODO: Inspect error for invalid cipher error or proxy host resolution failure.
+		return unexpected, err
+	}
+	tcpChan := make(chan error, 1)
```

Comment:
Nit: I would prefer `make(chan error)`, since you don't need it to be buffered.

Reply:
Done.